lib/spack/spack/util/path.py

```diff
@@ -320,40 +321,59 @@ def substitute_path_variables(path):
-    return os.path.sep.join(reps_list)
-
-
-def add_padding(path, length):
-    """Add padding subdirectories to path until total is length characters
-
-    Returns the padded path. If path is length - 1 or more characters long,
-    returns path. If path is length - 1 characters, warns that it is not
-    padding to length
-
-    Assumes path does not have a trailing path separator"""
-    padding_length = length - len(path)
-    if padding_length == 1:
-        # The only 1 character addition we can make to a path is `/`
-        # Spack internally runs normpath, so `foo/` will be reduced to `foo`
-        # Even if we removed this behavior from Spack, the user could normalize
-        # the path, removing the additional `/`.
-        # Because we can't expect one character of padding to show up in the
-        # resulting binaries, we warn the user and do not pad by a single char
-        tty.warn("Cannot pad path by exactly one character.")
-    if padding_length <= 0:
+def add_padding(path: str, length: int, fsencode: Callable[[str], bytes] = os.fsencode):
+    """Given a path ``path``, add placeholder subdirectories such that its length in code
+    units (depenending on the fs encoding) *without* trailing directory seperators is at
+    least ``length``.
+
+    This should work with UTF-16 encoding of paths, and also deal properly with paths containing
+    code points that are multiple code units. In principle since PEP 529 we could assume code
+    units are one byte, but unfortunately UTF-8 paths are not strictly enforced on Windows.
+
+    Returns: padded path as string, without trailing separator.
+    """
+
+    # Drop trailing dir seperators
+    path = path.rstrip(os.path.sep)
+
+    # NOTE: the assumption is that "." is one unit
+    bytes_per_unit = len(fsencode("."))
+
+    # How many units is our current path, and how many should we add.
+    units_in_path = len(fsencode(path)) // bytes_per_unit
+    pad_units = length - units_in_path
+
+    # If we can only fit a dir seperator, then we pad instead with _.
+    if pad_units <= 0:
         return path
 
-    # we subtract 1 from the padding_length to account for the path separator
-    # coming from os.path.join below
-    padding = _get_padding_string(padding_length - 1)
+    # os.path.sep should really be one unit, but let's be generic.
+    units_per_sep = len(fsencode(os.path.sep)) // bytes_per_unit
+    if pad_units <= units_per_sep:
+        return os.path.join(path, "_")
+
+    # Our padding including dir separator is how many units?
+    sep_with_dirname = os.path.sep + SPACK_PATH_PADDING_CHARS
+    bytes_per_sep_with_dirname = len(fsencode(sep_with_dirname))
+
+    # Ensure that there are no multi-unit characters (this way it's valid to truncate on any unit)
+    assert bytes_per_sep_with_dirname == len(sep_with_dirname) * bytes_per_unit
+    units_per_sep_with_dirname = bytes_per_sep_with_dirname // bytes_per_unit
```

Comment:
If we're focused on units instead of bytes, can't we do

```suggestion
    # Our padding including dir separator is how many units?
    sep_with_dirname = os.path.sep + SPACK_PATH_PADDING_CHARS
    units_per_sep_with_dirname = len(sep_with_dirname)
```

since we can truncate on any unit anyway if the measure we're using is units?